File: 98.validate-binary-search-tree.cpp
// @before-stub-for-debug-begin
#include <vector>
#include <string>
#include "commoncppproblem98.h"

using namespace std;
// @before-stub-for-debug-end

/*
 * @lc app=leetcode id=98 lang=cpp
 *
 * [98] Validate Binary Search Tree
 *
 * https://leetcode.com/problems/validate-binary-search-tree/description/
 *
 * algorithms
 * Medium (30.66%)
 * Likes:    9933
 * Dislikes: 896
 * Total Accepted:    1.4M
 * Total Submissions: 4.7M
 * Testcase Example:  '[2,1,3]'
 *
 * Given the root of a binary tree, determine if it is a valid binary search
 * tree (BST).
 *
 * A valid BST is defined as follows:
 *
 *
 * The left subtree of a node contains only nodes with keys less than the
 * node's key.
 * The right subtree of a node contains only nodes with keys greater than the
 * node's key.
 * Both the left and right subtrees must also be binary search trees.
 *
 *
 *
 * Example 1:
 *
 *
 * Input: root = [2,1,3]
 * Output: true
 *
 *
 * Example 2:
 *
 *
 * Input: root = [5,1,4,null,null,3,6]
 * Output: false
 * Explanation: The root node's value is 5 but its right child's value is
 * 4.
 *
 *
 *
 * Constraints:
 *
 *
 * The number of nodes in the tree is in the range [1, 10^4].
 * -2^31 <= Node.val <= 2^31 - 1
 *
 *
 */

 // @lc code=start
 /**
  * Definition for a binary tree node.
  * struct TreeNode {
  *     int val;
  *     TreeNode *left;
  *     TreeNode *right;
  *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
  *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
  * };
  */
class Solution {
private:
    bool isValidBST(TreeNode* node, long low, long up) {
        if (node == NULL)
            return true;
        if (node->val <= low || node->val >= up)
            return false;

        return isValidBST(node->left, low, node->val) && isValidBST(node->right, node->val, up);
    }

public:
    bool isValidBST(TreeNode* root) {
        return isValidBST(root, LONG_MIN, LONG_MAX);
    }
};
// @lc code=end

File: Leetcode-Solutions/98.validate-binary-search-tree.java
import java.util.Stack;

import javax.swing.tree.TreeNode;

/*
 * @lc app=leetcode id=98 lang=java
 *
 * [98] Validate Binary Search Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode() {} TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) { this.val = val; this.left
 * = left; this.right = right; } }
 */
class Solution {
    /*
     * public boolean isValidBST(TreeNode root) {
     * Stack<TreeNode> stack = new Stack<>();
     * TreeNode pre = null;
     * 
     * while (root != null || !stack.isEmpty()) {
     * while (root != null) {
     * stack.push(root);
     * root = root.left;
     * }
     * root = stack.pop();
     * if (pre != null && root.val <= pre.val) {
     * return false;
     * }
     * pre = root;
     * root = root.right;
     * }
     * return true;
     * }
     * 
     */

    private boolean isValidBST(long below, long up, TreeNode node) {
        if (node == null)
            return true;
        else if (node.val >= up || node.val <= below)
            return false;
        else
            return isValidBST(below, node.val, node.left) && isValidBST(node.val, up, node.right);
    }

    public boolean isValidBST(TreeNode root) {
        return isValidBST(Long.MIN_VALUE, Long.MAX_VALUE, root);
    }

}
// @lc code=end
